### PR TITLE
Add IO::Stapled

### PR DESCRIPTION
**Pull request by straight-shoota**
Adds `IO::Stapled` class which staples together two unidirectional `IO`s to form a single bidirectional `IO`. It basically delegates all read methods to `reader` and all write methods to `writer`.

This class is an extension to unidirectional IOs (like `IO::Memory`, `IO.pipe`) when you need true bidirectional IOs, for example as a replacement for socket streams.

This is particularly useful for speccing code that would normally work with a socket. You don't need to create an actual socket but can just staple together two `IO::Memory`.

### CODE REVIEW

**Comment by straight-shoota** [comment]
Added `gets` delegator and removed win32 check :+1:

File: src/io/stapled.cr [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          yield a, b
+        ensure
+          a.close
+          b.close
+        end
+      end
+    end
+  end
+
+  # Creates a pair of bidirectional pipe endpints connected with each other
```
Comment (jhass) [inline src/io/stapled.cr]: "endpoints"

File: src/io/stapled.cr [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# This class staples together two unidirectional `IO`s to form a single,
+# bidirectional `IO`.
+#
+# Example (loopback):
```
Comment (asterite) [inline src/io/stapled.cr]: I would use another example, maybe using two `IO::Memory`. When I first saw this I thought "why not use `IO.pipe` directly?".

**Comment by ysbaddaden** [comment]
Is there any use case other than "mocking sockets in specs"?

**Comment by straight-shoota** [comment]
@ysbaddaden You could for example wrap the standard streams with HTTP or any other protocol that expects read and write on the same IO ;)

**Comment by asterite** [comment]
@ysbaddaden I think the idea is to use it for specs, to stub a socket. But if we are going to do that, maybe `IO::Stapled` should just have two `IO::Memory` internally, and maybe be called `IO::Mock` or something like that.

**Comment by straight-shoota** [comment]
Yes, the main purpose of this is obviously mocking socket connections, but it can also be generally useful for multiplexing two unidirectional read/write streams in one IO.

Limiting to `IO::Memory` won't do because you can't really communicate through it. This would limit the usefulness as replacement for socket connections. For many use cases you'll need a pipe (hence there is also a class method to set that up).

**Comment by asterite** [comment]
Hm, maybe wrapping two pipes is a good use case...

**Pull request merged**